prow/pod-utils/gcs/options.go

```diff
@@ -0,0 +1,197 @@
+	fs.BoolVar(&o.DryRun, "dry-run", true, "do not interact with GCS")
+	return &o
+}
+
+// Options exposes the configuration necessary
+// for defining where in GCS an upload will land.
+type Options struct {
+	// Items are files or directories to upload
+	Items []string
+
+	// SubDir is appended to the resolved GCS path
+	SubDir string
+
+	// PathStrategy and the default Org and Repo
+	// determine how the GCS path is resolved from
+	// Prow-specified environment variables
+	PathStrategy string
+	DefaultOrg   string
+	DefaultRepo  string
+
+	GcsBucket          string
+	GceCredentialsFile string
+	DryRun             bool
+}
+
+func (o *Options) Complete(args []string) {
+	o.Items = args
+}
+
+func (o *Options) Validate() error {
+	if !o.DryRun {
+		if o.GcsBucket == "" {
+			return errors.New("GCS upload was requested but required flag --gcs-bucket was unset")
+		}
+
+		if o.GceCredentialsFile == "" {
+			return errors.New("GCS upload was requested but required flag --gcs-credentials-file was unset")
+		}
+	}
+
+	strategy := pathStrategyType(o.PathStrategy)
+	if strategy != pathStrategyLegacy && strategy != pathStrategyExplicit && strategy != pathStrategySingle {
+		return fmt.Errorf("GCS path strategy must be one of %q, %q, or %q", pathStrategyLegacy, pathStrategyExplicit, pathStrategySingle)
+	}
+
+	if strategy != pathStrategyExplicit && (o.DefaultOrg == "" || o.DefaultRepo == "") {
+		return fmt.Errorf("default org and repo must be provided for GCS strategy %q", strategy)
+	}
+
+	return nil
+}
+
+// Run will upload files to GCS as prescribed by
+// the options. Any extra files can be passed as
+// a parameter and will have the prefix prepended
+// to their destination in GCS, so the caller can
+// operate relative to the base of the GCS dir.
+func (o *Options) Run(extra map[string]UploadFunc) error {
+	var builder RepoPathBuilder
+	switch pathStrategyType(o.PathStrategy) {
+	case pathStrategyExplicit:
+		builder = NewExplicitRepoPathBuilder()
+	case pathStrategyLegacy:
+		builder = NewLegacyRepoPathBuilder(o.DefaultOrg, o.DefaultRepo)
+	case pathStrategySingle:
+		builder = NewSingleDefaultRepoPathBuilder(o.DefaultOrg, o.DefaultRepo)
+	}
+
+	spec, err := pjutil.ResolveSpecFromEnv()
+	if err != nil {
+		return fmt.Errorf("could not resolve job spec: %v", err)
+	}
+
+	var gcsPath string
+	jobBasePath := PathForSpec(spec, builder)
+	if o.SubDir == "" {
+		gcsPath = jobBasePath
+	} else {
+		gcsPath = path.Join(jobBasePath, o.SubDir)
+	}
+
+	uploadTargets := map[string]UploadFunc{}
+
+	// ensure that an alias exists for any
+	// job we're uploading artifacts for
+	if alias := AliasForSpec(spec); alias != "" {
+		uploadTargets[alias] = DataUpload(strings.NewReader(jobBasePath))
+	}
+
+	for _, item := range o.Items {
+		info, err := os.Stat(item)
+		if err != nil {
+			logrus.Warnf("Encountered error in resolving items to upload for %s: %v", item, err)
+			continue
+		}
+		if info.IsDir() {
+			gatherArtifacts(item, gcsPath, uploadTargets)
```

Comment:
Due to the way that the paths get calculated we do upload contents of a directory `artifacts/` under `<gcsPath>/artifacts`.

Reply:
Maybe I am missing something, but that doesn't seem to be the case? The upload path that is used in `gatherArtifacts` is `path.Join(gcsPath, filepath.Rel(artifactDir, fspath))`. The relative path won't include `artifacts` because it is relative to `artifacts`.

Reply:
Sorry, you're right -- we are one level up in my jobs:

```
time="2018-02-20T21:58:13Z" level=info msg="Gathering artifacts from artifact directory: /data/gcs/" 
time="2018-02-20T21:58:13Z" level=info msg="Found /data/gcs/artifacts/generated/avc_denials.log in artifact directory. Uploading as artifacts/generated/avc_denials.log" 
```

Do you think the `--sub-dir` option gives you the knobs you need for this? Or you think we need to treat `--sub-dir a/ foo/bar.txt` as an upload to `<gcsPath>/a/foo/bar.txt` and just `foo/bar.txt` as an upload to `<gcsPath>/foo/bar.txt`?

Reply:
The `--sub-dir` option might give us all the knobs we need, but it seems unnecessarily confusing to drop the directory name of artifacts directories. Unless there is a benefit besides shortening the upload path by one word I think it would be better to upload artifacts to their relative paths (i.e. upload the `artifacts` directory as is to `gcsPath` instead of removing the `artifacts` part of each artifact path).
This would eliminate any potential for name collisions and make it more obvious where artifacts are coming from.